### DISCUSSION TITLE
How To Pose Map?

### DISCUSSION THREAD

**Discussion by jdbott** [acmerobotics/road-runner-quickstart#381] [Q&A]
On the roadrunner website (https://rr.brott.dev/docs/v1-0/guides/pose-mapping/), I've found information about applying pose maps to trajectories. I was wondering if there is a way to make a method that takes in an Action, applies the pose map, and then returns a new Action with those changes applied. If anyone knows how, let me know!

Thanks!

**Comment by rbrott** [comment]
> I was wondering if there is a way to make a method that takes in an Action, applies the pose map, and then returns a new Action with those changes applied.

There is not because the `Action` interface has no concept of poses. Of course you can do structural recursion over the concrete action types in your codebase and apply the pose map when necessary. (By structural recursion, I mean a recursive method that takes an action, does an `instanceof` check for each relevant subtype, casts, and calls itself on any subactions. And if you end up pose mapping a trajectory, make doubly sure that it won't violate the constraints.)

I would instead recommend making a method that takes a pose map and creates an action.

**Reply by jdbott** [reply]
Thank you so much for the suggestion! I've been trying to implement this for a few hours now...it's not going too well. I'm not exactly sure how to create even the initial framework! Could you help me a bit to start off?

Thanks!

**Reply by rbrott** [reply]
Do you have an example opmode that I can demonstrate with?

**Reply by jdbott** [reply]
Sure! Here's a basic LinearOpMode with an Action I would like to apply a pose map to. (i.e. flip across the x-axis)

package org.firstinspires.ftc.teamcode;

import com.acmerobotics.roadrunner.Action;
import com.acmerobotics.roadrunner.Pose2d;
import com.acmerobotics.roadrunner.Vector2d;
import com.acmerobotics.roadrunner.ftc.Actions;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;

import org.firstinspires.ftc.teamcode.RRstuff.MecanumDriveJasonChassis;

@TeleOp
public class TrajTest extends LinearOpMode {

    @Override
    public void runOpMode() {
        MecanumDriveJasonChassis drive = new MecanumDriveJasonChassis(hardwareMap, new Pose2d(60, -58.5, Math.toRadians(90)));

        Action traj;

        traj = drive.actionBuilder(drive.pose)
                .setTangent(Math.toRadians(90))
                .splineTo(new Vector2d(22, -16), Math.toRadians(180))
                .setTangent(Math.toRadians(180))
                .splineTo(new Vector2d(-1, -55), Math.toRadians(180))
                .setTangent(Math.toRadians(180))
                .splineTo(new Vector2d(-28, -16), Math.toRadians(180))
                .setTangent(Math.toRadians(180))
                .splineTo(new Vector2d(-52,-55), Math.toRadians(-90))
                .strafeToLinearHeading(new Vector2d(60, -50), Math.toRadians(90))
                .build();

        waitForStart();

        if (isStopRequested()) return;

        Actions.runBlocking(
                traj
        );
    }
}
